sdk/core/Azure.Core/tests/common/ModelReaderWriter/Models/DiscriminatorSet/BaseModel.cs

```diff
@@ -0,0 +1,137 @@
+using System.ClientModel;
+using System.ClientModel.Primitives;
+using System.Text.Json;
+using Azure.Core.Tests.Common;
+
+namespace Azure.Core.Tests.ModelReaderWriterTests.Models
+{
+    [PersistableModelProxy(typeof(UnknownBaseModel))]
+    public abstract class BaseModel : IUtf8JsonSerializable, IJsonModel<BaseModel>
+    {
+        private Dictionary<string, BinaryData> _rawData;
+
+        protected internal BaseModel(Dictionary<string, BinaryData> rawData)
+        {
+            _rawData = rawData ?? new Dictionary<string, BinaryData>();
+        }
+
+        public string Kind { get; internal set; }
+        public string Name { get; set; }
+
+        protected internal void SerializeRawData(Utf8JsonWriter writer)
+        {
+            //write out the raw data
+            foreach (var property in _rawData)
+            {
+                writer.WritePropertyName(property.Key);
+#if NET6_0_OR_GREATER
+                writer.WriteRawValue(property.Value);
+#else
+                    JsonSerializer.Serialize(writer, JsonDocument.Parse(property.Value.ToString()).RootElement);
+#endif
+            }
+        }
+
+        void IUtf8JsonSerializable.Write(Utf8JsonWriter writer) => ((IJsonModel<BaseModel>)this).Write(writer, ModelReaderWriterHelper.WireOptions);
+
+        void IJsonModel<BaseModel>.Write(Utf8JsonWriter writer, ModelReaderWriterOptions options)
+        {
+            ModelReaderWriterHelper.ValidateFormat(this, options.Format);
+
+            Serialize(writer, options);
+        }
+
+        private void Serialize(Utf8JsonWriter writer, ModelReaderWriterOptions options)
+        {
+            writer.WriteStartObject();
+            writer.WritePropertyName("kind"u8);
+            writer.WriteStringValue(Kind);
+            if (Optional.IsDefined(Name))
+            {
+                writer.WritePropertyName("name"u8);
+                writer.WriteStringValue(Name);
+            }
+            if (options.Format == "J")
+            {
+                SerializeRawData(writer);
+            }
+            writer.WriteEndObject();
+        }
+
+        internal static BaseModel DeserializeBaseModel(BinaryData data, ModelReaderWriterOptions options)
+            => DeserializeBaseModel(JsonDocument.Parse(data.ToString()).RootElement, options);
+
+        internal static BaseModel DeserializeBaseModel(JsonElement element, ModelReaderWriterOptions options = default)
+        {
+            options ??= ModelReaderWriterHelper.WireOptions;
+
+            if (element.ValueKind == JsonValueKind.Null)
+            {
+                return null;
+            }
+            if (element.TryGetProperty("kind", out JsonElement discriminator))
+            {
+                switch (discriminator.GetString())
+                {
+                    case "X":
+                        return ModelX.DeserializeModelX(element, options);
+                    case "Y":
+                        return ModelY.DeserializeModelY(element, options);
```

Comment:
The default case happens below the case statement since we won't hit any return statements here https://github.com/Azure/azure-sdk-for-net/pull/39478/files#diff-9fd2d3becdb5ecd896f7b0716ddd023fda4f5556cc42605f6658e241fa5199a7R88-R110